docs/modules/record-validation/proc-record-validation.adoc

```diff
@@ -54,7 +54,14 @@ Replace the token `<rule definition>`  in the YAML configuration with either a S
 
 .Example Schema Validation Rule Definition
 
-The Schema Validation rule validates that the key or value matches the identified schema.
+The Schema Validation rule validates that the key or value matches a schema identified by its global ID within an Apicurio Schema Registry.
+
+If the key or value does not adhere to the schema, the record will be rejected.  
+
+Additionally, if the kafka producer has embedded a global ID within the record it will be validated against the global ID defined by the rule. If they do not match, the record will be rejected.  See the 
+{apicurio-docs}/getting-started/assembly-using-kafka-client-serdes.html#_consumer_schema_configuration[Apicurio documentation^] for details
+on how the global ID could be embedded into the record. 
+The filter supports extracting ID's from either the Apicurio `globalId` record header or from the initial bytes of the serialized content itself.
```

Comment:
```suggestion
The filter supports extracting ID's from either the default Apicurio `globalId` record headers or from the initial bytes of the serialized content itself.
```

the record could have a key header and a value header.

Reply:
>the record could have a key header and a value header. 

Urgh. What happens if it has both? Presumably one takes precedence? Should we, in a future PR, validate that they both point to the same schema?

Reply:
it depends if we are validating a key or a value.

So if we have config:

```
  - type: RecordValidation
    config:
        rules:
        - topicNames:
            - "topicA"
          keyRule:
            schemaValidationConfig:
              apicurioGlobalId: 1001
              apicurioRegistryUrl: http://registry.local:8080
            allowNulls: true
            allowEmpty: true
          valueRule:
            schemaValidationConfig:
              apicurioGlobalId: 1002
              apicurioRegistryUrl: http://registry.local:8080
            allowNulls: true
            allowEmpty: true
```

Then we require the key to be JSON with schema 1001 and the value to be JSON with schema 1002

When we are checking the key, we will look for the header `apicurio.key.globalId`. If that's not present we look for the magic byte at the start of the key's content and if we find it we load the 8-byte id from the key. If we find a global id then we will compare it to the configured one as an additional check, before proceeding to schema validation. If we didn't locate an id, we proceed straight to the schema validation.

The process for the value is the same except we look for header `apicurio.value.globalId` and look into the value contents when looking for magic-byte + id.

Maybe worth adding an IT as the existing one is all in terms of `valueRule`.

Reply:
This is possible because at the lowest level out BytebufValidator is aware of whether it's validating a key or value:

`CompletionStage<Result> validate(ByteBuffer buffer, Record record, boolean isKey);`

Reply:
If I'm following you we use the header if we see it and ignore the value of the magic byte, but does that actually work? Doesn't the parser need to know to skip the magic byte if its there? 

And to go back to my original question what should the filter do if header says 1001 and the magic byte says 1002 ?

Reply:
gotcha I thought you were asking about header precedence.

our code prefers headers, and doesn't read from the buffer if a global id is extracted from the headers. So it would fail to decode the JSON if header and magic-byte were somehow used together.

apicurio libs do it the other way around, they check for magic byte first. If present, decode id, else look for headers.

I believe via the apicurio producer you can't combine them, would have to check though.

Reply:
AHh I realise now I had miss read https://github.com/kroxylicious/kroxylicious/pull/1435#discussion_r1721067918 it wasn't suggesting both header and magic byte but header and value schemas.

Reply:
ah cool, is that cleared up then? The value and key use distinct header names so one shouldn't impact the other.

Reply:
Yes 😁